Double primitives and wrappers are not deserialised property, instead they are empty String, followed by NumberFormatException

Hi all,

I left message for the Jackson folks as I am having issues now with the deserialiser after upgrading Jackson databind..
Here is the discussion - https://github.com/FasterXML/jackson-databind/discussions/3918

After upgrading from `2.11.3` to `2.15.0`, we are now having issues on the deserialisation where double values are deliberately marshalled as empty string. 

Few areas that we already covered: 
  1. Data is stored as double in the Database prior retrieval
  2. Bean Deserialiser has mapped our class and sees double field as double (primitive)
  3. JsonLocation is `[Source: (String); byte offset: #UNKNOWN]` (version that works is `[Source: (String)"type"; line: -1, column: -1]`)
  4. Token in `VALUE_NUMBER_FLOAT `
  5. Disabled `MapperFeature.ALLOW_COERCION_OF_SCALARS`, same error I'm afraid
  6. `objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)`, no difference at all
  7. JsonSetter(nulls=Null.AS_EMPTY), no difference at all
  8. Don't see a need for a CustomDeserializer as it was working in previous versions and my POJO is simple


Error Log
```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: empty String (through reference chain: uk.co.***.ui.insession.CachedBiometric["biometric"]->uk.co.***.ui.insession.BehaviouralBiometric["metrics"]->java.util.ArrayList[0]->uk.co.***.ui.insession.MetricGroupDTO["metrics"]->java.util.ArrayList[0]->uk.co.***.ui.insession.metrics.ContinuousMultiSessionMetric["value"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:402)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:361)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1830)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:570)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:439)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1409)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:220)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:170)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:136)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:361)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:545)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:568)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:439)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1409)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:545)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:568)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:439)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1409)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:545)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:568)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:439)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1409)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:2079)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1229)
	at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:43)
	... 109 common frames omitted
Caused by: java.lang.NumberFormatException: empty String
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)

```

.ContinuousMultiSessionMetric class is a subtype of MultiSessionMetric. Here's how they are declared

```java
@JsonTypeInfo(use = Id.NAME, property = "type", include = As.EXISTING_PROPERTY)
@JsonSubTypes({
    @JsonSubTypes.Type(value = ContinuousMultiSessionMetric.class, name = "continuous"),
    @JsonSubTypes.Type(value = DiscreteMultiSessionMetric.class, name = "discrete")
})
public abstract class MultiSessionMetricDTO {
    @JsonProperty
    public final String name;

    @JsonProperty
    public final String desc;

    @JsonProperty
    public final String type;

    @JsonProperty
    public double normalisedValue;
...
}
```

```java
@JsonTypeName("continuous")
public class ContinuousMultiSessionMetric extends MultiSessionMetricDTO {
    @JsonProperty
    public final double value;
    @JsonProperty
    public final double mean;
    @JsonProperty
    public final double standardDeviation;

@JsonCreator
    ContinuousMultiSessionMetric(
            @JsonProperty("value") double value,
            @JsonProperty("mean") double mean,
            @JsonProperty("standardDeviation") double standardDeviation,
            @JsonProperty("name") String name,
            @JsonProperty("desc") String desc,
            @JsonProperty("normalisedValue") double normalisedValue
    ) {
        super(name, desc, "continuous", normalisedValue);
        this.value = value;
        this.mean = mean;
        this.standardDeviation = standardDeviation;
    }
```

```java
@JsonTypeName("discrete")
public class DiscreteMultiSessionMetric extends MultiSessionMetricDTO {

    @JsonProperty("value")
    public final String value;
    @JsonProperty("mode")
    public final String mode;

    @JsonCreator
    public DiscreteMultiSessionMetric(
        @JsonProperty("value") String value,
        @JsonProperty("mode") String mode,
        @JsonProperty("name") String name,
        @JsonProperty("desc") String desc,
        @JsonProperty("normalisedValue") double normalisedValue
    ) {
        super(name, desc, "discrete", normalisedValue);
        this.mode = mode;
        this.value = value;
    }
```

```json response from Mongo
{
    "_id" : ObjectId("6455799cc5631e7c84edf0a9"),
    "lastSystemEventId" : "EventID1",
    "sessionId" : "session1",
    "biometric" : {
        "metrics" : [
            {
                "metrics" : [
                    {
                        "value" : 2299.9,
                        "mean" : 2299.9,
                        "standardDeviation" : 0.0,
                        "name" : "Event/Received Time Difference",
                        "desc" : "The time difference between events happening and processing them for this session",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : "Chrome",
                        "mode" : "Chrome",
                        "name" : "Browser",
                        "desc" : "Browser used for this session",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    },
                    {
                        "value" : "MacOS",
                        "mode" : "MacOS",
                        "name" : "OS",
                        "desc" : "OS used for this session",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    },
                    {
                        "name" : "Country",
                        "desc" : "Country in which this session took place",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    },
                    {
                        "value" : "en-GB",
                        "mode" : "en-GB",
                        "name" : "Localisation",
                        "desc" : "Locale setting of the computer used for this session, which is a set of parameters defining language, region, etc",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    },
                    {
                        "value" : "1440x900",
                        "mode" : "1440x900",
                        "name" : "Screen Resolution",
                        "desc" : "Screen resolution of the computer used for this session",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    }
                ],
                "name" : "System Features",
                "desc" : "These features relate to the computer used for the session."
            },
            {
                "metrics" : [
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Click Duration",
                        "desc" : "Average time duration of a mouse click",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Double Click Latency",
                        "desc" : "Average time interval between double left clicks",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Left Clicks",
                        "desc" : "Total number of left clicks",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Right Clicks",
                        "desc" : "Total number of right clicks",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    }
                ],
                "name" : "Mouse Features",
                "desc" : "These features relate to the mouse use of the user during the session."
            },
            {
                "metrics" : [
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Typing Speed",
                        "desc" : "Average typing speed",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Key Press Duration",
                        "desc" : "Average time duration when a key is pressed down",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Key Latency",
                        "desc" : "Average time gap between the release of the last key and press of the next key",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Key Overlap Time",
                        "desc" : "Total duration of the time when the next key or keys are pressed down while the previous key has not been released",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Double Key Press Latency",
                        "desc" : "Average time interval between the same key being double pressed",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Keyboard Idle Time Percentage",
                        "desc" : "The percentage of time with no keyboard input",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Tab Presses",
                        "desc" : "Total number of Tab presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Left Shift Presses",
                        "desc" : "Total number of left shift presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Right Shift Presses",
                        "desc" : "Total number of right shift presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Caps Lock Presses",
                        "desc" : "Total number of Caps Lock presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Digits",
                        "desc" : "Total number of presses of the digit keys at the top of the alphanumeric section in keyboard",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Numberpad",
                        "desc" : "Total number of numpad key presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Delete Presses",
                        "desc" : "Total number of Delete key presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Backspaces",
                        "desc" : "Total number of Backspace key presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    }
                ],
                "name" : "Keyboard Features",
                "desc" : "These features relate to the keyboard use of the user during the session."
            },
            {
                "metrics" : [
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Idle Time Percentage",
                        "desc" : "The percentage of time with no click or keyboard input",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Average Time Between Text Fields",
                        "desc" : "Average time interval between the last keyboard input in a text field and the first one in the next text field",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Autofill Uses",
                        "desc" : "The number times Autofill has been used to fill out a form",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Autofill Cancels",
                        "desc" : "The number times Autofill has been canceled",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    }
                ],
                "name" : "Miscellaneous Features",
                "desc" : "Miscellaneous other features of the user."
            }
        ],
        "contexts" : [
            {
                "name" : "checkout",
                "alias" : "checkout"
            }
        ]
    },
    "context" : "checkout",
    "tenantId" : "TENANT",
    "entityId" : "consumer1",
    "entityType" : "consumer",
    "channel" : "JSC",
    "isGlobal" : false,
    "global" : false
}
{
    "_id" : ObjectId("6455799fc5631e7c84edf0aa"),
    "lastSystemEventId" : "EventID2",
    "sessionId" : "session2",
    "biometric" : {
        "metrics" : [
            {
                "metrics" : [
                    {
                        "value" : "UNKNOWN",
                        "mode" : "UNKNOWN",
                        "name" : "Charging",
                        "desc" : "The charging status during the session",
                        "normalisedValue" : 0.0,
                        "type" : "discrete"
                    }
                ],
                "name" : "Device Features",
                "desc" : "These features relate to the mobile device used for the session"
            },
            {
                "metrics" : [
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Keystroke Distribution",
                        "desc" : "The distribution of times between individual keystrokes",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Fraction of Time Typing",
                        "desc" : "The fraction of the total session time which was spent typing",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Typing Speed",
                        "desc" : "Average typing speed",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Time in Text Fields",
                        "desc" : "The total time spent typing in text fields",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    },
                    {
                        "value" : 8888.8,
                        "mean" : 8888.8,
                        "standardDeviation" : 0.0,
                        "name" : "Backspaces",
                        "desc" : "Total number of Backspace key presses",
                        "normalisedValue" : 0.0,
                        "type" : "continuous"
                    }
                ],
                "name" : "Typing Features",
                "desc" : "These features relate to the typing behaviour of the user"
            },
            {
                "metrics" : [
                    {
                        "value" : "-0.0",
                        "mode" : "0.0",
                        "name" : "Field Filling Sequence",
                        "desc" : "The sequence in which the user interacted with text fields",
                        "normalisedValue" : -0.0,
                        "type" : "discrete"
                    },
                    {
                        "value" : "-0.0",
                        "mode" : "0.0",
                        "name" : "View Sequence",
                        "desc" : "The sequence in which the user visited views within the app",
                        "normalisedValue" : -0.0,
                        "type" : "discrete"
                    }
                ],
                "name" : "Navigation Features",
                "desc" : "These features relate to the sequence in which the user navigates elements of the app"
            }
        ],
        "contexts" : [
            {
                "name" : "checkout",
                "alias" : "checkout"
            }
        ]
    },
    "context" : "checkout",
    "tenantId" : "TENANT",
    "entityId" : "consumer1",
    "entityType" : "consumer",
    "channel" : "IOS",
    "isGlobal" : false,
    "global" : false
}

```
So, this is how data are displayed when using Jackson:2.10.0
<img width="937" alt="Screenshot 2023-05-05 at 23 31 22" src="https://user-images.githubusercontent.com/6883670/236579031-c49ff348-954d-4687-9cca-1c92bc29fe52.png">

... And this is how it is displayed after upgrading
<img width="1280" alt="Screenshot 2023-05-05 at 23 32 34" src="https://user-images.githubusercontent.com/6883670/236579554-ac3adc8c-67e8-46af-8ef5-113825c44257.png">

Any help will be massively appreciate here.

Cheers

@rlodge I created a repository with a failing test. https://github.com/bernd/mongojack-jackson-2.15-issue

The tests run successfully with Jackson 2.14.3 but fail when changing the version to 2.15.2 (or 2.15.0).

The issue seems to be related to subtypes. A simple test with double primitives works, but in the subtypes test case it fails.

<details>
<summary>Maven output</summary>

```
[INFO] --- surefire:3.1.2:test (default-test) @ mongojack-jackson-2.15-issue ---
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.example.test.SimpleTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.195 s -- in org.example.test.SimpleTest
[INFO] Running org.example.test.JsonSubtypesTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.790 s <<< FAILURE! -- in org.example.test.JsonSubtypesTest
[ERROR] org.example.test.JsonSubtypesTest.test -- Time elapsed: 3.785 s <<< ERROR!
java.lang.RuntimeException: IOException encountered while parsing
	at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:45)
	at org.mongojack.internal.stream.JacksonCodec.decode(JacksonCodec.java:71)
	at com.mongodb.internal.operation.CommandResultArrayCodec.decode(CommandResultArrayCodec.java:52)
	at com.mongodb.internal.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:60)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:87)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:42)
	at org.bson.internal.LazyCodec.decode(LazyCodec.java:53)
	at org.bson.codecs.BsonDocumentCodec.readValue(BsonDocumentCodec.java:104)
	at com.mongodb.internal.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:63)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:87)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:42)
	at com.mongodb.internal.connection.ReplyMessage.<init>(ReplyMessage.java:50)
	at com.mongodb.internal.connection.InternalStreamConnection.getCommandResult(InternalStreamConnection.java:538)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:423)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:340)
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:116)
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:643)
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:71)
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:206)
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:119)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:85)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:75)
	at com.mongodb.internal.connection.DefaultServer$OperationCountTrackingConnection.command(DefaultServer.java:293)
	at com.mongodb.internal.operation.CommandOperationHelper.createReadCommandAndExecute(CommandOperationHelper.java:233)
	at com.mongodb.internal.operation.FindOperation.lambda$execute$1(FindOperation.java:332)
	at com.mongodb.internal.operation.OperationHelper.lambda$withSourceAndConnection$0(OperationHelper.java:356)
	at com.mongodb.internal.operation.OperationHelper.withSuppliedResource(OperationHelper.java:381)
	at com.mongodb.internal.operation.OperationHelper.lambda$withSourceAndConnection$1(OperationHelper.java:355)
	at com.mongodb.internal.operation.OperationHelper.withSuppliedResource(OperationHelper.java:381)
	at com.mongodb.internal.operation.OperationHelper.withSourceAndConnection(OperationHelper.java:354)
	at com.mongodb.internal.operation.FindOperation.lambda$execute$2(FindOperation.java:329)
	at com.mongodb.internal.async.function.RetryingSyncSupplier.get(RetryingSyncSupplier.java:67)
	at com.mongodb.internal.operation.FindOperation.execute(FindOperation.java:340)
	at com.mongodb.internal.operation.FindOperation.execute(FindOperation.java:79)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:191)
	at com.mongodb.client.internal.FindIterableImpl.first(FindIterableImpl.java:213)
	at org.mongojack.internal.util.FindIterableDecorator.first(FindIterableDecorator.java:291)
	at org.example.test.JsonSubtypesTest.test(JsonSubtypesTest.java:36)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: empty String (through reference chain: org.example.test.JsonSubtypesTest$GrumpyPerson["politeness"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:402)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:361)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1853)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:572)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:439)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1409)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:220)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:170)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:136)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:2079)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1229)
	at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:43)
	... 40 more
Caused by: java.lang.NumberFormatException: empty String
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:651)
	at com.fasterxml.jackson.core.io.NumberInput.parseDouble(NumberInput.java:388)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getNumberValue(TokenBuffer.java:1936)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getNumberValue(TokenBuffer.java:1891)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getDoubleValue(TokenBuffer.java:1842)
	at com.fasterxml.jackson.core.util.JsonParserDelegate.getDoubleValue(JsonParserDelegate.java:209)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$DoubleDeserializer.deserialize(NumberDeserializers.java:679)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$DoubleDeserializer.deserialize(NumberDeserializers.java:663)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:545)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:570)
	... 53 more

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   JsonSubtypesTest.test:36 » Runtime IOException encountered while parsing
[INFO] 
[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10.953 s
[INFO] Finished at: 2023-09-12T16:49:18+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project mongojack-jackson-2.15-issue: 
[ERROR] 
[ERROR] Please refer to /home/bernd/test/mongojack-jackson-2.15-issue/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

</summary>

@rlodge I created a test case for this repository in https://github.com/mongojack/mongojack/pull/240.

The test runs with the current Jackson version in the pom.xml file. Changing the version to 2.15.2 makes the test fail.

<details>
<summary>Maven test output</summary>

```
$ mvn clean test -Dtest=TestIssue236
[...]
[INFO] --- surefire:2.22.0:test (default-test) @ mongojack ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[mongod output]2023-09-12T17:20:52.019+0200 I CONTROL  [main] note: noprealloc may hurt performance in many applications
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] MongoDB starting : pid=105622 port=12345 dbpath=/tmp/embedmongo-db-1784b950-915a-4790-ba9e-cfc2257001e1 64-bit host=t14
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] db version v3.6.5
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] git version: a20ecd3e3a174162052ff99913bc2ca9a839d618
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] allocator: tcmalloc
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] modules: none
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] build environment:
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten]     distarch: x86_64
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten]     target_arch: x86_64
[mongod output] 2023-09-12T17:20:52.023+0200 I CONTROL  [initandlisten] options: { net: { port: 12345 }, security: { authorization: "disabled" }, storage: { dbPath: "/tmp/embedmongo-db-1784b950-915a-4790-ba9e-cfc2257001e1", journal: { enabled: false }, mmapv1: { preallocDataFiles: false, smallFiles: true }, syncPeriodSecs: 0.0 } }
[mongod output] 2023-09-12T17:20:52.023+0200 I STORAGE  [initandlisten] 
[mongod output] 2023-09-12T17:20:52.023+0200 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
[mongod output] 2023-09-12T17:20:52.023+0200 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
[mongod output] 2023-09-12T17:20:52.023+0200 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15360M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),cache_cursors=false,log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),,log=(enabled=false),
[mongod output] 2023-09-12T17:20:52.526+0200 I STORAGE  [initandlisten] WiredTiger message [1694532052:526406][105622:0x7f4bec2689c0], txn-recover: Set global recovery timestamp: 0
[mongod output] 2023-09-12T17:20:52.626+0200 W STORAGE  [initandlisten] Detected configuration for non-active storage engine mmapv1 when current storage engine is wiredTiger
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] 
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] ** WARNING: This server is bound to localhost.
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] **          Remote systems will be unable to connect to this server. 
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] **          Start the server with --bind_ip <address> to specify which IP 
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] **          addresses it should serve responses from, or with --bind_ip_all to
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] **          bind to all interfaces. If this behavior is desired, start the
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] **          server with --bind_ip 127.0.0.1 to disable this warning.
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] 
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] 
[mongod output] 2023-09-12T17:20:52.626+0200 I CONTROL  [initandlisten] ** WARNING: soft rlimits too low. rlimits set to 126469 processes, 1048576 files. Number of processes should be at least 524288 : 0.5 times number of files.
[mongod output] 2023-09-12T17:20:52.626+0200 I STORAGE  [initandlisten] createCollection: admin.system.version with provided UUID: 4a227803-6cb2-4e11-92d7-8060c255772d
[mongod output] 2023-09-12T17:20:52.668+0200 I COMMAND  [initandlisten] setting featureCompatibilityVersion to 3.6
[mongod output] 2023-09-12T17:20:52.670+0200 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 1a0fbbd6-5c16-4540-a4d6-600bbc8738ed
[mongod output] 2023-09-12T17:20:52.717+0200 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/tmp/embedmongo-db-1784b950-915a-4790-ba9e-cfc2257001e1/diagnostic.data'
[mongod output] 2023-09-12T17:20:52.718+0200 I NETWORK  [initandlisten] waiting for connections on port 12345
[mongod output] [INFO] Running org.mongojack.TestIssue236
2023-09-12T17:20:52.855+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44330 #1 (1 connection now open)
[mongod output] 2023-09-12T17:20:52.855+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44328 #2 (2 connections now open)
[mongod output] 2023-09-12T17:20:52.862+0200 I NETWORK  [conn1] received client metadata from 127.0.0.1:44330 conn1: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:52.863+0200 I NETWORK  [conn2] received client metadata from 127.0.0.1:44328 conn2: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:53.132+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44342 #3 (3 connections now open)
[mongod output] 2023-09-12T17:20:53.133+0200 I NETWORK  [conn3] received client metadata from 127.0.0.1:44342 conn3: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:53.153+0200 I STORAGE  [conn3] createCollection: unittest.KqmtqGgP with generated UUID: 16eae537-4141-4cc1-8971-df94aaeddad0
[mongod output] 2023-09-12T17:20:53.241+0200 I COMMAND  [conn3] CMD: drop unittest.KqmtqGgP
[mongod output] 2023-09-12T17:20:53.241+0200 I STORAGE  [conn3] Finishing collection drop for unittest.KqmtqGgP (16eae537-4141-4cc1-8971-df94aaeddad0).
[mongod output] 2023-09-12T17:20:53.260+0200 I COMMAND  [conn3] CMD: drop unittest.FjgRMlKa
[mongod output] 2023-09-12T17:20:53.265+0200 I NETWORK  [conn3] end connection 127.0.0.1:44342 (2 connections now open)
[mongod output] 2023-09-12T17:20:53.265+0200 I NETWORK  [conn1] end connection 127.0.0.1:44330 (1 connection now open)
[mongod output] 2023-09-12T17:20:53.265+0200 I NETWORK  [conn2] end connection 127.0.0.1:44328 (0 connections now open)
[mongod output] 2023-09-12T17:20:53.273+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44356 #4 (1 connection now open)
[mongod output] 2023-09-12T17:20:53.273+0200 I NETWORK  [conn4] received client metadata from 127.0.0.1:44356 conn4: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:53.273+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44364 #5 (2 connections now open)
[mongod output] 2023-09-12T17:20:53.274+0200 I NETWORK  [conn5] received client metadata from 127.0.0.1:44364 conn5: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:53.279+0200 I NETWORK  [listener] connection accepted from 127.0.0.1:44372 #6 (3 connections now open)
[mongod output] 2023-09-12T17:20:53.279+0200 I NETWORK  [conn6] received client metadata from 127.0.0.1:44372 conn6: { driver: { name: "mongo-java-driver|sync", version: "4.8.1" }, os: { type: "Linux", name: "Linux", architecture: "amd64", version: "6.2.0-32-generic" }, platform: "Java/Private Build/17.0.8.1+1-Ubuntu-0ubuntu122.04" }
[mongod output] 2023-09-12T17:20:53.282+0200 I STORAGE  [conn6] createCollection: unittest.LIMpwShm with generated UUID: a8fe4677-83ec-4471-ba34-15cdc3740e49
[mongod output] 2023-09-12T17:20:53.327+0200 I COMMAND  [conn6] CMD: drop unittest.LIMpwShm
[mongod output] 2023-09-12T17:20:53.327+0200 I STORAGE  [conn6] Finishing collection drop for unittest.LIMpwShm (a8fe4677-83ec-4471-ba34-15cdc3740e49).
[mongod output] 2023-09-12T17:20:53.342+0200 I COMMAND  [conn6] CMD: drop unittest.oWnHflBz
[mongod output] 2023-09-12T17:20:53.343+0200 I NETWORK  [conn6] end connection 127.0.0.1:44372 (2 connections now open)
[mongod output] 2023-09-12T17:20:53.343+0200 I NETWORK  [conn4] end connection 127.0.0.1:44356 (1 connection now open)
[mongod output] 2023-09-12T17:20:53.343+0200 I NETWORK  [conn5] end connection 127.0.0.1:44364 (0 connections now open)
[mongod output] [ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.623 s <<< FAILURE! - in org.mongojack.TestIssue236
[ERROR] testSubtypes(org.mongojack.TestIssue236)  Time elapsed: 0.501 s  <<< ERROR!
java.lang.RuntimeException: IOException encountered while parsing
	at org.mongojack.TestIssue236.testSubtypes(TestIssue236.java:60)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: empty String (through reference chain: org.mongojack.TestIssue236$GrumpyPerson["politeness"])
	at org.mongojack.TestIssue236.testSubtypes(TestIssue236.java:60)
Caused by: java.lang.NumberFormatException: empty String
	at org.mongojack.TestIssue236.testSubtypes(TestIssue236.java:60)


2023-09-12T17:20:53.447+0200 I CONTROL  [signalProcessingThread] got signal 2 (Interrupt), will terminate after current cmd ends
[mongod output] 2023-09-12T17:20:53.447+0200 I NETWORK  [signalProcessingThread] shutdown: going to close listening sockets...
[mongod output] 2023-09-12T17:20:53.447+0200 I NETWORK  [signalProcessingThread] removing socket file: /tmp/mongodb-12345.sock
[mongod output] 2023-09-12T17:20:53.447+0200 I FTDC     [signalProcessingThread] Shutting down full-time diagnostic data capture
[mongod output] 2023-09-12T17:20:53.449+0200 I STORAGE  [signalProcessingThread] WiredTigerKVEngine shutting down
[mongod output] 2023-09-12T17:20:53.469+0200 I STORAGE  [signalProcessingThread] shutdown: removing fs lock...
[mongod output] 2023-09-12T17:20:53.469+0200 I CONTROL  [signalProcessingThread] now exiting
[mongod output] 2023-09-12T17:20:53.469+0200 I CONTROL  [signalProcessingThread] shutting down with code:0
[mongod output] 
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   TestIssue236.testSubtypes:60 » Runtime IOException encountered while parsing
[INFO] 
[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  7.485 s
[INFO] Finished at: 2023-09-12T17:20:53+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project mongojack: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/bernd/Code/src/mongojack/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

</details>

There is a hint by the Jackson maintainer that might be helpful: https://github.com/FasterXML/jackson-databind/discussions/3918#discussioncomment-5876332

--- a/pom.xml
+++ b/pom.xml
@@ -20,7 +20,7 @@
       <dependency>
         <groupId>com.fasterxml.jackson</groupId>
         <artifactId>jackson-bom</artifactId>
-        <version>2.14.1</version>
+        <version>2.15.2</version>
         <scope>import</scope>
         <type>pom</type>
       </dependency>

--- a/src/main/java/org/mongojack/internal/stream/JsonParserAdapter.java
+++ b/src/main/java/org/mongojack/internal/stream/JsonParserAdapter.java
@@ -277,6 +277,11 @@ public class JsonParserAdapter extends ParserBase {
     }
 
     @Override
+    public Object getNumberValueDeferred() throws IOException {
+        return getNumberValue();
+    }
+
+    @Override
     public JsonParser.NumberType getNumberType() {
         if (currentValue == null) {
             return null;

--- a/src/test/java/org/mongojack/TestParsingAndGeneratingPolymorphicTypes.java
+++ b/src/test/java/org/mongojack/TestParsingAndGeneratingPolymorphicTypes.java
@@ -1,0 +1,102 @@
+package org.mongojack;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mongodb.client.model.Filters;
+import org.junit.Before;
+import org.junit.Test;
+import org.mongojack.internal.MongoJackModule;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.junit.MatcherAssert.assertThat;
+
+/**
+ * Test parsing and generating polymorphic types.
+ * <p>Related issues:
+ *   <ul>
+ *     <li><a href="https://github.com/mongojack/mongojack/issues/236">mongojack/mongojack#236</a></li>
+ *   </ul>
+ * </p>
+ */
+public class TestParsingAndGeneratingPolymorphicTypes extends MongoDBTestBase {
+    private JacksonMongoCollection<Metric> coll;
+    private ObjectMapper objectMapper;
+
+    @Before
+    public void setup() throws Exception {
+        objectMapper = MongoJackModule.configure(new ObjectMapper());
+        coll = getCollection(Metric.class, objectMapper);
+    }
+
+    @Test
+    public void testSubtypes() throws Exception {
+        final Metric longMetric = new LongMetric(236);
+        final Metric doubleMetric = new DoubleMetric(0.8);
+
+        // Validate that the serialize/deserialize cycle works when using plain Jackson
+        assertThat(objectMapper.readValue(objectMapper.writeValueAsString(longMetric), Metric.class), instanceOf(LongMetric.class));
+        assertThat(objectMapper.readValue(objectMapper.writeValueAsString(doubleMetric), Metric.class), instanceOf(DoubleMetric.class));
+
+        coll.insertOne(longMetric);
+        coll.insertOne(doubleMetric);
+
+        final Metric longMetricRecord = coll.findOne(Filters.eq("type", "long"));
+        final Metric doubleMetricRecord = coll.findOne(Filters.eq("type", "double"));
+
+        assertThat(longMetricRecord, instanceOf(LongMetric.class));
+        assertThat(((LongMetric) longMetricRecord).getValue(), equalTo(236L));
+        assertThat(doubleMetricRecord, instanceOf(DoubleMetric.class));
+        assertThat(((DoubleMetric) doubleMetricRecord).getValue(), equalTo(0.8d));
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "type")
+    @JsonSubTypes({
+        @JsonSubTypes.Type(value = LongMetric.class, name = "long"),
+        @JsonSubTypes.Type(value = DoubleMetric.class, name = "double"),
+    })
+    @JsonIgnoreProperties("_id")
+    public static abstract class Metric {
+        @JsonProperty("type")
+        public abstract String getType();
+    }
+
+    public static class LongMetric extends Metric {
+        private final long value;
+
+        public LongMetric(@JsonProperty("value") long value) {
+            this.value = value;
+        }
+
+        @JsonProperty("value")
+        public long getValue() {
+            return value;
+        }
+
+        @Override
+        public String getType() {
+            return "long";
+        }
+    }
+
+    public static class DoubleMetric extends Metric {
+        private final double value;
+
+        public DoubleMetric(@JsonProperty("value") double value) {
+            this.value = value;
+        }
+
+        @JsonProperty("value")
+        public double getValue() {
+            return value;
+        }
+
+        @Override
+        public String getType() {
+            return "double";
+        }
+    }
+}

